packages/sync/src/utils/withDeferredSyncSafeguards.ts

```diff
@@ -0,0 +1,41 @@
+import { BaseDatabase } from '@tupaia/database';
+
+import { switchTombstoneTriggers } from './switchTombstoneTriggers';
+
+export const withDeferredSyncSafeguards = async <T>(
+  database: BaseDatabase,
+  operation: () => Promise<T>,
+): Promise<T> => {
+  /**
+   * Defer foreign key constraint assertions until the end of the transaction.
+   *
+   * This prevents constraint violations during data synchronization when dealing with
+   * self-referencing foreign keys (e.g., entity.parent_id, permission_group.parent_id).
+   *
+   * Without deferral, these constraints would be checked immediately upon insertion,
+   * causing failures when parent records haven't been inserted yet. The hierarchical
+   * nature of this data makes it complicated to guarantee correct insertion order.
+   *
+   * Note: Only constraints explicitly altered to be DEFERRABLE will be affected.
+   * Standard foreign key constraints continue to be enforced immediately.
+   */
+  await database.executeSql(`
+    SET CONSTRAINTS ALL DEFERRED;
+  `);
```

Comment:
This is also an important point worth highlighting here, all explained in the comment.

It's definitely feasible to sort records within a single table to respect hierarchical dependencies, but it will introduce significant complexity and performance overhead.

Reply:
Really really clever idea!!